text/0000-module-resolution.md

```diff
@@ -0,0 +1,340 @@
+
+### Types of ModuleRecord
+
+#### AliasModuleRecord
+
+This type of resolution maps a given module specifier to a particular path within the root of that module.
+
+```json
+{
+    "modules": [
+       {
+            "name": "ui/button",
+            "path": "src/modules/ui/button/button.js"
+        }        
+    ]
+}
+```
+
+`AliasModuleRecord` must contain `name` and a `path` keys.
+
+#### DirectoryModuleRecord
+
+This type of resolution allows to point to to a folder that contain LWC modules.
+Its important to note that within that folder the structure must be very specific, it must contain a namespace folder, then a folder per module bundle (moduleName) adnd them the moduleEntry file that matches the name of the bundle (the entry must be of type .css, .html, .js or .ts)
+
+```json
+{
+    "modules": [
+       { "dir" : "src/modules" }
+    ]
+}
+```
+
+Here is an example of a valid module resolution given the previous DirectoryModuleRecord
+```
+src
+└── modules/ 
+    └── ui/ (namespace)
+        └── button/ (moduleName)
+            └── button.js (moduleEntry)
+```
+
+`DirectoryModuleRecord` must contain `dir` key.
+
+#### NpmModuleRecord
+
+This type of resolution tells the resolver to find an npm package with that name and resolve the modules it might have inside. This process can be recursive (see example at the top).
+
+```json
+{
+    "modules": [
+       { "npm" : "@ui/components" }     
+    ]
+}
+```
+
+`NpmModuleRecord` must contain `npm` key.
+
+## Algorithm
+
+Here is the algoritm for module resolution (the instructions are not in a rigurous specification format for brevity and time)
+
+```
+  Let `rootDir` be the initial path for the application, `initModules` 
+  the list of ModuleRecords to be resolved and ModuleRecordEntryList 
+  an empty array.
+  
+  1. Load the LWC configuration file in the rootDir
+    1.1 Search for `lwc.config.js`, if it exist read it.
+    1.2 If lwc.config.js is not found search in package.json
+    (If both are found the configuration file has precedence).
+
+  2. Merge all the ModuleRecords found in the config with 
+  the list of initModules. If an alias with the same name 
+  exist pick initModules over the ones resolved in the config.
+
+  3. For each ModuleRecord:
+
+    3.1  Validate that its a valid ModuleRecord type, skip it otherwise.
+
+    3.2 Resolve ModuleRecord entry:
+
+        3.2.1 If is an AliasModuleRecord validate the path and add create 
+        a ModuleRecordEntry with: 
+          - `specifier` as the `name` value.
+          - `scope` with the closest configuration path.
+          - `entry` with the `path` value.
+
+        3.2.2 If is a DirModuleRecord validate the path and find of the 
+        modules that match the structure:
+        [namespace]/[componentName]/[componentName.{html|css|js|ts}]
+
+        3.2.3 If is a NpmModuleRecord, validate that the package is installed 
```

Comment:
> validate that the package is installed

By looking in `node_modules`, correct?

Also, we should add that we should validate that it is a bare module specifier and has no leading `/` or `./`.

Reply:
Yes that is done in 3.1. I will add some steps there